# Практическая работа №2: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 1304 Андреев Вячеслав. Вариант №27.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа, привить навык использования
систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения


### Аддитивная цепочка

Аддитивной цепочкой натурального числа $n$ называется последовательность натуральных чисел<br> 
$$1 = a_0, a_1, a_2,\dots,a_r = n,$$
где
$$a_i = a_j + a_k,\quad \forall k \leq j < i,\quad i = 1, 2, \dots, r.$$<br>

$r$ - длина цепочки<br>
$l(n) = r$ - длина кратчайшей цепочки для заданного $n$<br>

Оценки для $l(n)$:<br>
$$l(n) \leq \lambda(n) +\nu(n) - 1$$
$$l(n) \geq \lceil \log_{2}(n) \rceil$$

Пара $(j, k), \; 0 \leqslant k \leqslant j < i$ - называется шагом.<br>

Виды шагов:
- Удвоение: $j = k = i - 1$
- Звёздный: $j = i - 1$
- Малый: $\lambda(a_i) = \lambda(a_{i-1})$

Свойства шагов:
- Первый шаг всегда удвоение.
- Удвоение - звёздный шаг, но никогда не малый шаг.
- За удвоением всегда следует звёздный шаг.
- Если $i$-ый шаг не малый, то $(i+1)$ шаг либо малый, либо звёздный, либо и тот, и другой.
- Если $(i+1)$ шаг ни звездный, ни малый, то $i$-ый шаг должен быть малым.

Аддитивная звёздная цепочка:<br>
- Включает в себя только звёздные шаги
- Длина звездной цепочки $l^*(n)$
- Шаг звёздной цепочки $a_i = a_{i-1} + a_k \; \forall k < i$
- $l(n) \leqslant l^*(n)$


### Алгоритм Брауэра

- Количество умножений в ходе работы алгоритма: 
$$log_2(n) = \lambda(n) + \frac{\lambda(n)}{\lambda(\lambda(n))} + O\left(\frac{\lambda(n) cdot\lambda(\lambda(\lambda(n)))}{(\lambda(\lambda(n)))^2}\right)$$
- Самая короткая аддитивная цепчока для числа $n$ имеет длину не более $\lambda(n)$.
- Доказано, что почти для всех $n$ минимальная аддитивная цепчока имеет длину $log_2(n)$.

Для наутральных $n$ при заданном натуральном числе $k$ цепочка может быть построена с помощью реккурентной формулы:
$$B_k(n) =   \begin{cases}
1, 2, 3,\dots, 2^k-1       & \quad \text{если } n < 2^k\\
B_k(q), 2q, 4q, 8q,\dots, 2^kq, n  & \quad \text{если } n \geqslant 2^k \text{ и } q = \lfloor\frac{n}{2^k}\rfloor\end{cases}$$<br>

Длина цепочки: $log_2(n) = j(k+1) + 2^k - 2$, если
$$jk \leqslant \lambda(n) < (j+1)k.$$
Для больших $n$ длина будет минимизирована, если
$$k = \lambda(\lambda(n)) - 2\lambda(\lambda(\lambda(n)))$$


Алгоритм:<br>
1. Задаётся фиксированный параметр $k$ для рассматриваемого числа $n$.\
Вычисление вспомогательных чисел:
$$d = 2^k, \quad q_1 = [\frac{n}{d}], \quad r_1 = n \; mod \; d \implies n = q_1d + r_1 \; (0 \leqslant r_1 < d) \\ q_2 = [\frac{q_1}{d}], \quad r_2 = q_1 mod \; d \implies q_1 = q_2d + r_2 \; (0 \leqslant r_2 < d)$$
2. Продолжаем до тех пор, пока не появится такое $q_s < d$:
$$q_{s-1} = d_sd + r_s.$$
3. $n$ имеет вид:
$$n = 2^kq_1 + r_1 = 2^k(2^kq_2 + r_2) + r_1 = \dots = 2^k(2^k(\dots(2^kq_s + r_s)\dots)+ r_2) + r_1$$


### Алгоритм дробления вектора индексов

Векторо индексов - последовательность $r_1,r_2,\dots,r_{m-1}$, где<br>
$r_i=\{z:1\leq z\leq i\}, \quad a_i=a_{i-1}+a_{r_{i-1}},\quad2\leq i\leq m-1.$

Пусть $r=\{r_i\}_{i=1}^{m-1}$, $\widetilde r=\{\widetilde r_i\}_{i=1}^{m-1}$.<br>
Тогда $r \succ \widetilde r$, если $r_1=\widetilde r_1,r_2=\widetilde r_2,\dots,r_{m-1}=\widetilde r_{m-1}$, а $r_m>\widetilde r_m$.<br>

Рассмотрим вектор индекса вида: $\{r_1, r_2, \dots, r_q\} \cup \{ \rho_{q+1}, \rho_{q+2},\dots, \rho_{m}\}.$<br>
Назовём левую часть фиксированной, а правую - меняющейся.<br>

Последний элемент цепочки $a_{max}=a_{q+1}*2^{m-q}$ при векторе индексов $\{r_1, r_2, \dots, r_q\}\cup\{q+1,\dots,m\}\quad$<br>
Последний элемент цепочки $a_{min}=a_{q+1}+m-q$ при векторе индексов $\{r_1, r_2, \dots, r_q\}\cup\{1,\dots,1\}\quad $<br>

Алгоритм:
<ol>
    <li>Цикл по длинам $\underline l(n) \le m \le \overline l(n)$, $q=\frac{m}{2}$ - индекс дроблений</li>
    <li>Цикл по всем фиксированным частям. На каждом шаге вычисляем $a_{min}, a_{max}$ и строим цепочку.
        <ol>
          <li>Если $a_{m} = n$, то задача решена.</li>
          <li>Если $n \notin [a_{min}, a_{max}]$, то переходим к следующему набору $\{r_1, r_2, \dots, r_q\}$</li>
          <li>Если $n \in [a_{min}, a_{max}]$, то организуем внутренний цикл перебора меняющейся части.
                    <ol>
          <li>Если $a_{m} = a_{n}$, то задача решена</li>
          <li>Иначе переходим к следующему (по введённой упорядоченности) фиксированной части $\{ \rho_{q+1}, \rho_{q+2},\dots, \rho_{m}\}$</li>
         </ol>
            </li>
        </ol>
    </li>
    <li>Если все наборы фиксированной длины исчерпаны, то увеличиваем их длину во внешнем цикле</li>
</ol>


### Гипотеза Шольца-Брауэра

$$l(2^{n}-1) \le l(n) + n - 1$$

- Гипотеза доказана для звёздных цепочек
- Гипотеза справедлива для всех $n < 578469$
- Равенство выполняется для всех $1 \leqslant n \leqslant 64$


## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать.

## Выполнение работы



### Реализация алгоритма Брауэра

Функция *BrauerSolve* принимает на вход:</br>
*n* - число, для которого необходимо найти аддитивную цепочку</br>
*k* - фиксированный параметр

In [98]:
def BrauerSolve(n, k):
    LN = n.digits(2^k) # convert n into 2^k base counting system
    LResult = [i for i in range(1, 2^k)] # fixed result start
    LResult.append(LN[-1])
    
    for i in range(len(LN) - 2, -1, -1):
        q = LResult[-1]
        for j in range(1, k + 1):
            LResult.append((2^j) * q)
        LResult.append(LResult[-1] + LN[i])   
    return sorted(set(LResult)) # remove copies and sort

Проверим работу алгоритма на 𝑛 = 7, 22, 55, 123, 669, 1000, 1555</br>
𝑘 = 2, 3, 4, 5, 6</br>
Сравним длину получившейся цепочки с длиной минимальной цепочки.

In [99]:
def TestBrauer():
    NTable = {
              7:[1, 2, 3, 6, 7], 
              22:[1, 2, 4, 5, 10, 11, 22], 
              55:[1, 2, 3, 6, 12, 13, 26, 52, 55],
              123:[1, 2, 4, 8, 16, 32, 40, 41, 82, 123], 
              669:[1, 2, 4, 5, 10, 20, 40, 41, 82, 83, 166, 167, 334, 668, 669], 
              1000:[1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000], 
              1555:[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 513, 1026, 1539, 1555]
             }
    
    for n in NTable:
        for k in range(2, 7):
            LResult = BrauerSolve(n, k)
            print(f"|{n}|{k}|{LResult}|{NTable[n]}|{len(LResult)}|{len(NTable[n])}|")
            
TestBrauer()

|7|2|[1, 2, 3, 4, 7]|[1, 2, 3, 6, 7]|5|5|
|7|3|[1, 2, 3, 4, 5, 6, 7]|[1, 2, 3, 6, 7]|7|5|
|7|4|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]|[1, 2, 3, 6, 7]|15|5|
|7|5|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]|[1, 2, 3, 6, 7]|31|5|
|7|6|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63]|[1, 2, 3, 6, 7]|63|5|
|22|2|[1, 2, 3, 4, 5, 10, 20, 22]|[1, 2, 4, 5, 10, 11, 22]|8|7|
|22|3|[1, 2, 3, 4, 5, 6, 7, 8, 16, 22]|[1, 2, 4, 5, 10, 11, 22]|10|7|
|22|4|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 22]|[1, 2, 4, 5, 10, 11, 22]|17|7|
|22|5|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]|[1, 2, 4, 5, 10, 11, 22]|31|7|
|22|6|[1, 2, 3, 4, 5, 6, 7, 8, 9,

|n|k|Цепочка Брауэра|Минимальная цепочка|Длина Брауэра|Длина минимальной|
|:----:|:-:|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------:|:-----------------------------------------------------:|:-------------:|:-----------------:|
|7|2|[1, 2, 3, 4, 7]|[1, 2, 3, 6, 7]|5|5|
|7|3|[1, 2, 3, 4, 5, 6, 7]|[1, 2, 3, 6, 7]|7|5|
|7|4|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]|[1, 2, 3, 6, 7]|15|5|
|7|5|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]|[1, 2, 3, 6, 7]|31|5|
|7|6|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63]|[1, 2, 3, 6, 7]|63|5|
|22|2|[1, 2, 3, 4, 5, 10, 20, 22]|[1, 2, 4, 5, 10, 11, 22]|8|7|
|22|3|[1, 2, 3, 4, 5, 6, 7, 8, 16, 22]|[1, 2, 4, 5, 10, 11, 22]|10|7|
|22|4|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 22]|[1, 2, 4, 5, 10, 11, 22]|17|7|
|22|5|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]|[1, 2, 4, 5, 10, 11, 22]|31|7|
|22|6|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63]|[1, 2, 4, 5, 10, 11, 22]|63|7|
|55|2|[1, 2, 3, 6, 12, 13, 26, 52, 55]|[1, 2, 3, 6, 12, 13, 26, 52, 55]|9|9|
|55|3|[1, 2, 3, 4, 5, 6, 7, 12, 24, 48, 55]|[1, 2, 3, 6, 12, 13, 26, 52, 55]|11|9|
|55|4|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 24, 48, 55]|[1, 2, 3, 6, 12, 13, 26, 52, 55]|18|9|
|55|5|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 55]|[1, 2, 3, 6, 12, 13, 26, 52, 55]|33|9|
|55|6|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63]|[1, 2, 3, 6, 12, 13, 26, 52, 55]|63|9|
|123|2|[1, 2, 3, 4, 7, 14, 28, 30, 60, 120, 123]|[1, 2, 4, 8, 16, 32, 40, 41, 82, 123]|11|10|
|123|3|[1, 2, 3, 4, 5, 6, 7, 8, 15, 30, 60, 120, 123]|[1, 2, 4, 8, 16, 32, 40, 41, 82, 123]|13|10|
|123|4|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 28, 56, 112, 123]|[1, 2, 4, 8, 16, 32, 40, 41, 82, 123]|19|10|
|123|5|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 48, 96, 123]|[1, 2, 4, 8, 16, 32, 40, 41, 82, 123]|34|10|
|123|6|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 123]|[1, 2, 4, 8, 16, 32, 40, 41, 82, 123]|65|10|
|669|2|[1, 2, 3, 4, 8, 10, 20, 40, 41, 82, 164, 167, 334, 668, 669]|[1, 2, 4, 5, 10, 20, 40, 41, 82, 83, 166, 167, 334, 668, 669]|15|15|
|669|3|[1, 2, 3, 4, 5, 6, 7, 8, 10, 20, 40, 80, 83, 166, 332, 664, 669]|[1, 2, 4, 5, 10, 20, 40, 41, 82, 83, 166, 167, 334, 668, 669]|17|15|
|669|4|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 41, 82, 164, 328, 656, 669]|[1, 2, 4, 5, 10, 20, 40, 41, 82, 83, 166, 167, 334, 668, 669]|23|15|
|669|5|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 40, 80, 160, 320, 640, 669]|[1, 2, 4, 5, 10, 20, 40, 41, 82, 83, 166, 167, 334, 668, 669]|37|15|
|669|6|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 80, 160, 320, 640, 669]|[1, 2, 4, 5, 10, 20, 40, 41, 82, 83, 166, 167, 334, 668, 669]|68|15|
|1000|2|[1, 2, 3, 6, 12, 15, 30, 60, 62, 124, 248, 250, 500, 1000]|[1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000]|14|13|
|1000|3|[1, 2, 3, 4, 5, 6, 7, 8, 15, 30, 60, 120, 125, 250, 500, 1000]|[1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000]|16|13|
|1000|4|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 24, 48, 62, 124, 248, 496, 992, 1000]|[1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000]|23|13|
|1000|5|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 62, 124, 248, 496, 992, 1000]|[1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000]|37|13|
|1000|6|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 120, 240, 480, 960, 1000]|[1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000]|68|13|
|1555|2|[1, 2, 3, 4, 6, 12, 24, 48, 96, 97, 194, 388, 776, 1552, 1555]|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 513, 1026, 1539, 1555]|15|14|
|1555|3|[1, 2, 3, 4, 5, 6, 7, 12, 24, 48, 96, 192, 194, 388, 776, 1552, 1555]|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 513, 1026, 1539, 1555]|17|14|
|1555|4|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 24, 48, 96, 97, 194, 388, 776, 1552, 1555]|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 513, 1026, 1539, 1555]|24|14|
|1555|5|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 48, 96, 192, 384, 768, 1536, 1555]|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 513, 1026, 1539, 1555]|39|14|
|1555|6|[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 96, 192, 384, 768, 1536, 1555]|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 513, 1026, 1539, 1555]|69|14|

#### Вывод
Реализован алгоритм *Брауэра* для нахождения аддитивной цепочки.<br>
Видно, что длина цепочки *Брауэра* может совпадать с длиной минимальной цепочки, но может быть больше неё. 
Таким образом, алгоритм *Брауэра* является приблизительным, но у него есть преимущество в скорости работы. 

### Реализация алгоритма дробления вектора индексов

Функция *IndexVectorToChain* - преобразовывает вектор инексов в аддитивную цепочку:</br>
*IndexVector* - вектор инексов для преобразования в аддитивную цепочку.

Функция *IsMinIndexVector* - проверяет не является ли вектор индексов минимальным:</br>
*IndexVector* - вектор для проверки, что он состоит только из единиц.

Функция *ReduceIndexVector* - уменьшает вектор индексов:</br>
*IndexVector* - вектор индексов для уменьшения.

Функция *IndexVectorSplitting* - основная функция реализации алгоритма дробления вектора индексов:</br>
*n* - число, для которого небходимо построить аддитивную цепочку.

In [100]:

def IndexVectorToChain(IndexVector): 
    LResult = [1]
    for i in IndexVector:
        LResult.append(LResult[-1] + LResult[i - 1])
    return LResult

def IsMinIndexVector(IndexVector):
    for i in IndexVector:
        if i > 1:
            return False
    return True

def ReduceIndexVector(IndexVector, q = 0):
    if IsMinIndexVector(IndexVector):
        return []
    for i in range(len(IndexVector) - 1, -1, -1):
        if IndexVector[i] > 1: 
            IndexVector[i] -= 1
            break  
        elif IndexVector[i] == 1:
            IndexVector[i] = i + q + 1
    return IndexVector

                
def IndexVectorSplitting(n):
    if n < 1:
        return []
    if n == 1: 
        return [1]
    
    LowerM = ceil(log(n, 2))
    UpperM = floor(log(n, 2)) + n.digits(2).count(1)
    for m in range(LowerM, UpperM + 1): 
        q = m // 2 + 1
        IndexVector = [i for i in range(1, m + 1)]
        ConstantPart = IndexVector[q:]
        DynamicPart = IndexVector[:q]
 
        while len(DynamicPart) > 0: 
            LChain = IndexVectorToChain(DynamicPart + ConstantPart)  
            AMin = LChain[q] + m - q
            AMax = LChain[q] * 2 ** (m - q)
            
            if LChain[-1] == n: 
                return LChain
            elif (n < AMin) or (n > AMax): 
                DynamicPart = ReduceIndexVector(DynamicPart)
            else:
                while len(ConstantPart) > 0:
                    LChain = IndexVectorToChain(DynamicPart + ConstantPart)
                    if n == LChain[-1]:
                        return LChain
                    ConstantPart = ReduceIndexVector(ConstantPart, q)
                DynamicPart = ReduceIndexVector(DynamicPart)
                ConstantPart = IndexVector[q:]

Проверим работу для n = 1008, 1025, 1030, 1100, 1555

In [101]:
import time
import multiprocessing as mp

def TestIndexVectorSplitting(n):
    TStart = time.time()
    LResult = IndexVectorSplitting(n)
    TEnd = time.time()
    return f"|{n}|{LResult}|{len(LResult)}|{round(TEnd - TStart, 3)}s|"
        

ThreadsPool = mp.Pool(mp.cpu_count())
LResults = ThreadsPool.map(TestIndexVectorSplitting, [n for n in [1008, 1025, 1030, 1100, 1555]])
ThreadsPool.close()

for r in LResults:
    print(r)

|1008|[1, 2, 4, 8, 16, 32, 64, 128, 256, 320, 336, 672, 1008]|13|10.665s|
|1025|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1025]|12|0.001s|
|1030|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1028, 1030]|13|8.234s|
|1100|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1088, 1096, 1100]|14|384.072s|
|1555|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 513, 1026, 1539, 1555]|14|96.544s|


|  n   |                             Цепочка                             | Длина цепочки |  Время работы  |
|:----:|:---------------------------------------------------------------:|:-------------:|:--------------:|
|1008|[1, 2, 4, 8, 16, 32, 64, 128, 256, 320, 336, 672, 1008]|13|10.788s|
|1025|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1025]|12|0.001s|
|1030|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1028, 1030]|13|8.088s|
|1100|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1088, 1096, 1100]|14|378.146s|
|1555|[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 513, 1026, 1539, 1555]|14|95.709s|

#### Вывод
Реализован алгоритм дробления вектора индексов.<br>
По сравнению с алгоритмом Брауэра, этот алгоритм работает намного дольше, но гарантированно строит минимальную звёздную цепочку. Повышенное время работы алгоритма обусловлено тем, что он является перебором.

### Проверка гипотезы Шольца-Брауэра

Проверим гипотезу Шольца Брауэра для звёздных цепочек, $n \le 12$.</br>
Оптимизируем алгоритм дробления вектора индексов. Перебирать будем только длины, равные $l(n) + n - 1$.

In [102]:
def IndexVectorSplittingNew(n, m):
    if n < 1:
        return []
    if n == 1: 
        return [1]
    
    m = m - 1
    q = m // 2 + 1
    IndexVector = [i for i in range(1, m + 1)]
    ConstantPart = IndexVector[q:]
    DynamicPart = IndexVector[:q]

    while len(DynamicPart) > 0: 
        LChain = IndexVectorToChain(DynamicPart + ConstantPart)  
        AMin = LChain[q] + m - q
        AMax = LChain[q] * 2 ** (m - q)

        if LChain[-1] == n: 
            return LChain
        elif (n < AMin) or (n > AMax): 
            DynamicPart = ReduceIndexVector(DynamicPart)
        else:
            while len(ConstantPart) > 0:
                LChain = IndexVectorToChain(DynamicPart + ConstantPart)
                if n == LChain[-1]:
                    return LChain
                ConstantPart = ReduceIndexVector(ConstantPart, q)
            DynamicPart = ReduceIndexVector(DynamicPart)
            ConstantPart = IndexVector[q:]

In [103]:
def ScholzBrauerCheck():
    for i in [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]:
        OriginalLen = len(IndexVectorSplitting(i)) + i - 1
        NewLen = len(IndexVectorSplittingNew(2^i - 1, OriginalLen))
        if OriginalLen != NewLen:
            print(f"|{i}|{NewLen}|{OriginalLen}|Incorrect|")
            break
        print(f"|{i}|{NewLen}|{OriginalLen}|Correct|")
        
ScholzBrauerCheck()

|1|1|1|Correct|
|2|3|3|Correct|
|3|5|5|Correct|
|4|6|6|Correct|
|5|8|8|Correct|
|6|9|9|Correct|
|7|11|11|Correct|
|8|11|11|Correct|
|9|13|13|Correct|
|10|14|14|Correct|
|11|16|16|Correct|
|12|16|16|Correct|


| n  |$$l(2^{n}-1)$$|$$l(n) + n - 1$$|    Результат   |
|:--:|--------------|----------------|:--------------:|
|1|1|1|Correct|
|2|3|3|Correct|
|3|5|5|Correct|
|4|6|6|Correct|
|5|8|8|Correct|
|6|9|9|Correct|
|7|11|11|Correct|
|8|11|11|Correct|
|9|13|13|Correct|
|10|14|14|Correct|
|11|16|16|Correct|
|12|16|16|Correct|

#### Вывод
Проверена гипотеза Шольца-Брауэра.</br>
Гипотеза верна для всех $n<=12$.

## Выводы

Сформировано представление об аддитивных цепочках, вырабатано умение составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа. Привит навык использования системы компьютерной математики SageMath для реализации алгоритмов.<br>
Реализованы алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа: алгоритм Брауэра и дробления вектора индексов. Произведено сравнение этих алгоритмов.<br>
С помощью алгоритма дробления вектора индексов проверена корректность гипотезы Шольца-Брауэра для $n: \; 1 \leqslant n \leqslant 12$.